upgrade_groupd.sh

```diff
@@ -0,0 +1,372 @@
+  namespace: template-service-broker
+spec:
+  channel: alpha
+  installPlanApproval: Automatic
+  name: templateservicebroker
+  source: installed-community-template-service-broker
+  sourceNamespace: template-service-broker
+  startingCSV: templateservicebrokeroperator.v0.2.0
+EOF
+  oc create -f tsbsubscription.yaml -n template-service-broker
+
+  echo "waiting for pods running....."
+  sleep 30
+
+cat << EOF > tsb-cr.yaml
+apiVersion: osb.openshift.io/v1alpha1
+kind: TemplateServiceBroker
+metadata:
+  name: template-service-broker
+  namespace: template-service-broker
+EOF
+  oc create -f tsb-cr.yaml -n template-service-broker
+
+  echo "waiting for pods running....."
+  sleep 30
+  if [ !$(oc get po -n template-service-broker | tail -n +2 |grep  -cvE "(Running|Completed)") ] ; then
+    echo -e  "\033[31m TSB install successfully, need a few minutes to wait tsb ready !  \033[0m"
+  else 
+    echo -e "\033[31m TSB installation failed , please manually install TSB on web console!  \033[0m"
+  fi
+}
+
+#DevExp team
+function devExp {
+  echo "Collectting pre_upgrade data for DevExp team.... "
+  echo -e "================== DEVEXP-TEAM  DATA `date` ============================================" >> $RESULT_FILE
+  echo -e "#oc get pods -o yaml -n openshift-image-registry | grep imageID |uniq\n$(oc get pods -o yaml -n openshift-image-registry | grep imageID |uniq)" >> $RESULT_FILE
+  echo -e "#oc get pods -o yaml -n openshift-cluster-samples-operator | grep imageID |uniq\n$(oc get pods -o yaml -n openshift-cluster-samples-operator | grep imageID |uniq)" >> $RESULT_FILE
+  echo -e "#oc get pods -n openshift-image-registry\n$(oc get pods -n openshift-image-registry)" >> $RESULT_FILE
+  echo -e "#oc describe is jenkins -n openshift\n$(oc describe is jenkins -n openshift)" >> $RESULT_FILE
+  echo -e "#oc describe is ruby -n openshift\n$(oc describe is ruby -n openshift)" >> $RESULT_FILE
+
+  echo "================== DEVEXP team's namespace ${DEVEXP_PRO} is created!"
+  oc new-project ${DEVEXP_PRO}
+  echo "=================Preparing upgrade data for DEVEXP team....."
+  oc new-app ruby~https://github.com/openshift/ruby-ex
+  oc new-app -f jenkins-ephemeral
+
+  echo "==================waiting for pods running....."
+  sleep 30
+  echo -e "#oc get all -n ${DEVEXP_PRO} \n$(oc get all -n ${DEVEXP_PRO})" >> $RESULT_FILE
+  echo "Data preparing for DEVEXP team was finished!"
+
+}
+
+#Monitoring
+function monitoring {
+  echo "Collectting pre_upgrade data for Monitoring team.... "
+  echo -e "================== MONITORING-TEAM  DATA `date` ============================================" >> $RESULT_FILE
+  echo  -e "#oc get clusteroperator monitoring -o yaml\n$(oc get clusteroperator monitoring -o yaml)" >> $RESULT_FILE
+  echo "=================Preparing upgrade data for MONITORING team....."
+  oc create -f https://raw.githubusercontent.com/juzhao/monitoring/master/config.yaml
+  echo "waiting for PVC creation....."
+  sleep 20
```

Comment:
I think it is not good to set a fixed sleep time, sometimes 20s is not enough, the following is better
PVC_total=`oc -n openshift-monitoring get pvc | grep -v NAME | wc -l`
while(( $PVC_total < 5 ))
do 
   echo "PVCs are creating"
done

Reply:
Agree.  Could we have a retry_until function? 
For example:
retry_until (time, command, result)
retry_until(60, "oc get elasticsearch --selector component=elasticsearch", "running")
retry_until(60, "es_cluster_healthy", "green")

Reply:
All these wait for services have already been implemented in `verification-tests`, that's why I'm suggesting to use the automation framework to do this prep-upgrade.